src/main/java/org/zalando/nakadi/service/subscription/zk/AbstractZkSubscriptionClient.java

```diff
@@ -200,12 +197,39 @@ public final void unregisterSession(final Session session) {
         }
 
         try {
+
+            final CountDownLatch latch = new CountDownLatch(zkSessions.size());
+            final Map<String, Session> resultSessions = new HashMap<>();
             for (final String sessionId : zkSessions) {
-                final int weight = Integer.parseInt(new String(getCurator().getData()
-                        .forPath(getSubscriptionPath("/sessions/" + sessionId)), UTF_8));
-                sessions.add(new Session(sessionId, weight));
+                getCurator().getData()
+                        .inBackground((client, event) -> {
+                            try {
+                                if (event.getResultCode() == KeeperException.Code.OK.intValue()) {
+                                    synchronized (resultSessions) {
+                                        resultSessions.put(
+                                                sessionId,
+                                                new Session(
+                                                        sessionId,
+                                                        Integer.parseInt(new String(event.getData(), UTF_8))));
```

Comment:
Can you please format this code in a better way? It's very difficult to read it as it has 28 nested levels.

Reply:
I tried to make it better